packages/web-components/fast-foundation/src/virtual-list/virtual-list.ts

```diff
@@ -0,0 +1,877 @@
+import {
+    attr,
+    DOM,
+    Notifier,
+    nullableNumberConverter,
+    Observable,
+    observable,
+    RepeatBehavior,
+    RepeatDirective,
+    Splice,
+    ViewTemplate,
+} from "@microsoft/fast-element";
+import { eventResize, eventScroll, Orientation } from "@microsoft/fast-web-utilities";
+import { FoundationElement } from "../foundation-element";
+import { IntersectionService } from "../utilities/intersection-service";
+import { IdleCallbackQueue } from "../utilities/idle-callback-queue";
+import type { ResizeObserverClassDefinition } from "../utilities/resize-observer";
+/**
+ * Defines when the component updates its position automatically.
+ *
+ * @public
+ */
+export type VirtualListAutoUpdateMode = "manual" | "viewport-resize" | "auto";
+
+/**
+ * Used to describe the position of an element within the list
+ *
+ * @public
+ */
+export interface SizeMap {
+    start: number;
+    end: number;
+    size: number;
+}
+
+/**
+ *  The VirtualList class
+ *
+ * @public
+ */
+export class VirtualList extends FoundationElement {
+    /**
+     *  Whether or not the display should virtualize
+     *
+     * @public
+     */
+    @attr({ attribute: "virtualization-enabled", mode: "boolean" })
+    public virtualizationEnabled: boolean = true;
+    private virtualizationEnabledChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.reset();
+        }
+    }
+
+    /**
+     * The HTML ID of the viewport element
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: anchor
+     */
+    @attr({ attribute: "viewport" })
+    public viewport: string = "";
+    private viewportChanged(): void {
+        if (this.$fastController.isConnected) {
+            this.viewportElement = this.getViewport();
+            this.updateDimensions();
+        }
+    }
+
+    /**
+     * The size in pixels of each item along the virtualization axis
+     *
+     * @public
+     * @remarks
+     * HTML Attribute: item-size
+     */
+    @attr({ attribute: "item-size", converter: nullableNumberConverter })
+    public itemSize: number = 50;
```

Comment:
Is there an issue with allowing this to support an "auto-size" notion? So, if this were set to NaN/null, I might expect (assuming `SizeMap` isn't set), that after the first render, you can retrieve the size of a rendered element, and simply use that as the `itemSize`. This API seems like it expects a consumer to know how big they want to render an element, which it seems common for a client to not know.

Reply:
The "price" of virtualization is that you have a quick way of calculating the size of things, so you don't want to have to do actual layout to figure that out.  I can see some scenarios where authors could have items have a default size that got updated when they actually render, like a feed that virtualized out of view elements as you scroll away, but I think we can leave that to implementors - ie. a component built on a virtual list that defines a default size map and updates sizes as elements load.  I think it would be a challenge to have a big list with arbitrary scrolling and dynamic element sizing not be janky or slow.

Reply:
Authors have a lot of control over the positioning of elements through the templates.  All my test "gallery-item-template" does is set the translateX based on the visible items data.  We don't actually enforce size/position of items beyond that.  Authors could pass down additional context data and have more elaborate transforms.  I could see building a transition like you describe, challenging but doable I think.

Reply:
Please only build something like I suggest if you feel personally motivated to do it. :) I certainly don't want to take any of your time to simply satisfy a curiosity of mine.